File: heavy-metal-bake-sale/typescript/src/BakeSale.ts
import { Money } from './Money.js';
import { Product } from './Product.js';
import {
  InsufficientPaymentException,
  OutOfStockException,
  UnknownPurchaseCodeException,
} from './Exceptions.js';

export class BakeSale {
  constructor(private readonly _inventory: Product[]) {}

  get inventory(): readonly Product[] {
    return this._inventory;
  }

  static createDefault(): BakeSale {
    return new BakeSale([
      new Product('Brownie', new Money(0.75), 'B', 48),
      new Product('Muffin', new Money(1.00), 'M', 36),
      new Product('Cake Pop', new Money(1.35), 'C', 24),
      new Product('Water', new Money(1.50), 'W', 30),
    ]);
  }

  calculateTotal(order: string): Money {
    const codes = order.split(',').map((c) => c.trim());
    const products = this.resolveProducts(codes);

    this.validateStock(products);

    let total = Money.zero;
    for (const product of products) {
      total = total.plus(product.price);
    }

    for (const product of products) {
      product.decrementStock();
    }

    return total;
  }

  calculateChange(total: Money, payment: Money): Money {
    if (payment.lessThan(total)) {
      throw new InsufficientPaymentException();
    }
    return payment.minus(total);
  }

  private resolveProducts(codes: string[]): Product[] {
    const products: Product[] = [];
    for (const code of codes) {
      const product = this._inventory.find((p) => p.purchaseCode === code);
      if (!product) {
        throw new UnknownPurchaseCodeException(code);
      }
      products.push(product);
    }
    return products;
  }

  private validateStock(products: Product[]): void {
    const grouped = new Map<string, { product: Product; count: number }>();
    for (const product of products) {
      const existing = grouped.get(product.purchaseCode);
      if (existing) {
        existing.count++;
      } else {
        grouped.set(product.purchaseCode, { product, count: 1 });
      }
    }
    for (const { product, count } of grouped.values()) {
      if (product.stock < count) {
        throw new OutOfStockException(product.name);
      }
    }
  }
}

File: heavy-metal-bake-sale/csharp/src/HeavyMetalBakeSale/BakeSale.cs
namespace HeavyMetalBakeSale;

public class BakeSale
{
    private readonly List<Product> _inventory;

    public BakeSale(List<Product> inventory)
    {
        _inventory = inventory;
    }

    public IReadOnlyList<Product> Inventory => _inventory;

    public static BakeSale CreateDefault()
    {
        return new BakeSale(new List<Product>
        {
            new("Brownie", new Money(0.75m), "B", 48),
            new("Muffin", new Money(1.00m), "M", 36),
            new("Cake Pop", new Money(1.35m), "C", 24),
            new("Water", new Money(1.50m), "W", 30),
        });
    }

    public Money CalculateTotal(string order)
    {
        var codes = order.Split(',').Select(c => c.Trim()).ToList();
        var products = ResolveProducts(codes);

        ValidateStock(products);

        var total = Money.Zero;
        foreach (var product in products)
        {
            total += product.Price;
        }

        foreach (var product in products)
        {
            product.DecrementStock();
        }

        return total;
    }

    public Money CalculateChange(Money total, Money payment)
    {
        if (payment < total)
        {
            throw new InsufficientPaymentException();
        }

        return payment - total;
    }

    private List<Product> ResolveProducts(List<string> codes)
    {
        var products = new List<Product>();
        foreach (var code in codes)
        {
            var product = _inventory.FirstOrDefault(p => p.PurchaseCode == code);
            if (product is null)
            {
                throw new UnknownPurchaseCodeException(code);
            }
            products.Add(product);
        }
        return products;
    }

    private static void ValidateStock(List<Product> products)
    {
        var grouped = products.GroupBy(p => p.PurchaseCode);
        foreach (var group in grouped)
        {
            var product = group.First();
            if (product.Stock < group.Count())
            {
                throw new OutOfStockException(product.Name);
            }
        }
    }
}
